api/ext/tor.py

```diff
@@ -1,11 +1,24 @@
 import ipaddress
 import os
-from collections import namedtuple  # TODO: dataclasses
+from dataclasses import dataclass
+from typing import Optional, Union
 
 from .. import settings
 
-HiddenService = namedtuple("HiddenService", ["name", "directory", "hostname", "port_definition"])
-PortDefinition = namedtuple("PortDefinition", ["virtual_port", "ip", "port"])
+
+@dataclass(frozen=True)
```

Comment:
What is the use of `frozen` here?
We don't need it to be immutable, plus, as from python docs:

> There is a tiny performance penalty when using frozen=True: __init__() cannot use simple assignment to initialize fields, and must use object.__setattr__().

https://docs.python.org/3/library/dataclasses.html#frozen-instances

Reply:
I used `frozen` as a "security" measure because the definition takes place only once and then the attributes of the dataclass doesn't change.

Reply:
Should I remove the `frozen=True` statement and make the dataclass mutable?

Reply:
No, thanks, good measure.